perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ConsentsManagerBlImpl.java

```diff
@@ -232,4 +234,28 @@ public void removeFacility(PerunSession sess, ConsentHub consentHub, Facility fa
 			getConsentsManagerImpl().deleteConsentHub(sess, consentHub);
 		}
 	}
+
+	@Override
+	public Consent changeConsentStatus(PerunSession sess, Consent consent, ConsentStatus status) throws ConsentNotExistsException, InvalidConsentStatusException {
+		if (status != ConsentStatus.GRANTED && status != ConsentStatus.REVOKED) {
+			throw new InvalidConsentStatusException("Invalid consent status value.");
+		}
+		if (status == consent.getStatus()) {
+			throw new InvalidConsentStatusException("Tried to set consent status on current value.");
+		}
+		consent.setStatus(status);
+
+		checkExistingConsents(sess, consent);
+		getConsentsManagerImpl().changeConsentStatus(sess, consent);
+		getPerunBl().getAuditer().log(sess, new ChangedConsentStatus(consent));
+		return getConsentById(sess, consent.getId());
+	}
+
+	private void checkExistingConsents(PerunSession sess, Consent consent) throws ConsentNotExistsException {
+		for (Consent currentConsent : getConsentsForUserAndConsentHub(sess, consent.getUserId(), consent.getConsentHub().getId())) {
+			if (currentConsent.getId() != consent.getId() && currentConsent.getStatus() != ConsentStatus.UNSIGNED) {
+				deleteConsent(sess, currentConsent);
```

Comment:
You could catch ConsentNotExistsException here and throw InternalErrorException/ConsitencyError, since theoretically the Consent should always exist. Then you could remove the exception from the method signature (which would make more sense, because you checked whether the consent exists in the entry layer).